linera-core/src/chain_worker/state/temporary_changes.rs

```diff
@@ -209,30 +208,19 @@ where
         self.0.chain.remove_bundles_from_inboxes(block).await?;
         // Verify that all required bytecode hashed certificate values and blobs are available, and no
         // unrelated ones provided.
+        let published_blob_ids = block.published_blob_ids();
         self.0
-            .check_no_missing_blobs(block.published_blob_ids(), blobs)
+            .check_no_missing_blobs(published_blob_ids.clone(), blobs)
             .await?;
         for blob in blobs {
+            Self::check_blob_size(blob.content(), &policy)?;
             self.0.cache_recent_blob(Cow::Borrowed(blob)).await;
         }
-        for blob in self.0.get_blobs(block.published_blob_ids()).await? {
-            let blob_size = match blob.content() {
-                BlobContent::Data(bytes) => bytes.len(),
-                BlobContent::ContractBytecode(compressed_bytecode)
-                | BlobContent::ServiceBytecode(compressed_bytecode) => {
-                    ensure!(
-                        compressed_bytecode.decompressed_size_at_most(maximum_bytecode_size)?,
-                        WorkerError::BytecodeTooLarge
-                    );
-                    compressed_bytecode.compressed_bytes.len()
-                }
-            };
-            ensure!(
-                u64::try_from(blob_size)
-                    .ok()
-                    .is_some_and(|size| size <= maximum_blob_size),
-                WorkerError::BlobTooLarge
-            )
+        let checked_blobs = blobs.iter().map(|blob| blob.id()).collect::<BTreeSet<_>>();
+        for blob in self.0.get_blobs(published_blob_ids).await? {
```

Comment:
If we already have these blobs (i.e. they're not part of `blobs` but `published_blob_ids`), shouldn't it be true that we've already checked their sizes against `policy`?

Reply:
Can't we make a check where only allowed blobs in `blobs` are the once referenced by `published_blob_ids`?

Reply:
We do! That's enforced in `get_missing_blobs`.